src/ccs.rs

```diff
@@ -0,0 +1,636 @@
+    let m = M.iter().fold(0, |acc, matrix| max(acc, matrix.n_rows()));
+    let n = M.iter().fold(0, |acc, matrix| max(acc, matrix.n_cols()));
+
+    let s = m.log_2() as usize;
+    let s_prime = n.log_2() as usize;
+
+    let shape = CCSShape {
+      num_cons,
+      num_vars,
+      num_io,
+      M,
+      t,
+      q,
+      d,
+      S,
+      c,
+      m,
+      n,
+      s,
+      s_prime,
+    };
+
+    Ok(shape)
+  }
+
+  // NOTE: Not using previous used multiply_vec (r1cs.rs), see utils.rs
+
+  // NOTE: Equivalent to is_sat_relaxed (r1cs.rs) but for CCCSS/LCCCS?
+  // Either here or as a separate method on LCCCS struct
+
+  /// Checks if the CCS instance is satisfiable given a witness and its shape
+  pub fn is_sat(
+    &self,
+    ck: &CommitmentKey<G>,
+    U: &CCSInstance<G>,
+    W: &CCSWitness<G>,
+  ) -> Result<(), NovaError> {
+    assert_eq!(W.W.len(), self.num_vars);
+    assert_eq!(U.X.len(), self.num_io);
+
+    // NOTE: All matricies have the same number of rows, but in a SparseMatrix we need to check all of them
+    // Can probably be made more efficient by keeping track fo n_rows/n_cols at creation/insert time
+    let m = self
+      .M
+      .iter()
+      .fold(0, |acc, matrix| max(acc, matrix.n_rows()));
+
+    // Sage code to check CCS relation:
+    //
+    // r = [F(0)] * m
+    // for i in range(0, q):
+    //     hadamard_output = [F(1)]*m
+    //     for j in S[i]:
+    //         hadamard_output = hadamard_product(hadamard_output,
+    //                 matrix_vector_product(M[j], z))
+    //
+    //     r = vec_add(r, vec_elem_mul(hadamard_output, c[i]))
+    // print("\nCCS relation check (∑ cᵢ ⋅ ◯ Mⱼ z == 0):", r == [0]*m)
+    //
+    // verify if ∑ cᵢ ⋅ ◯ Mⱼ z == 0
+    let res_eq: bool = {
+      let mut r = vec![G::Scalar::ZERO; m];
+      let z = concat(vec![W.W.clone(), vec![G::Scalar::ONE], U.X.clone()]);
+
+      for i in 0..self.q {
+        let mut hadamard_output = vec![G::Scalar::ONE; m];
+        for j in &self.S[i] {
+          let mvp = matrix_vector_product_sparse(&self.M[*j], &z)?;
+          hadamard_output = hadamard_product(&hadamard_output, &mvp)?;
+        }
+
+        let vep = vector_elem_product(&hadamard_output, &self.c[i])?;
+
+        r = vector_add(&r, &vep)?;
+      }
+      r == vec![G::Scalar::ZERO; m]
+    };
+
+    // verify if comm_W is a commitment to W
+    let res_comm: bool = U.comm_W == CE::<G>::commit(ck, &W.W);
+
+    if res_eq && res_comm {
+      Ok(())
+    } else {
+      Err(NovaError::UnSat)
+    }
+  }
+
+  pub fn from_r1cs(r1cs: R1CSShape<G>) -> Self {
+    // These contants are used for R1CS-to-CCS, see the paper for more details
+    const T: usize = 3;
+    const Q: usize = 2;
+    const D: usize = 2;
+    const S1: [usize; 2] = [0, 1];
+    const S2: [usize; 1] = [2];
+
+    let C0 = G::Scalar::ONE;
+    let C1 = -G::Scalar::ONE;
+
+    // NOTE: All matricies have the same number of rows, but in a SparseMatrix we need to check all of them
```

Comment:
Let's just create a fn that checks this impl for `SparseMatrix` so that we can remove the multiple annotations for this issue.